File: backend/routes/maintenance.py
"""
routes/maintenance.py - Gestion des fenetres de maintenance (calendrier de changements).

CRUD des fenetres + endpoint de verification (utilise par le frontend pour
prevenir avant une action mutante). L'enforcement reel est applique dans les
routes mutantes (updates, reboot) via maintenance.is_allowed().

Securite (OWASP) :
  - A01 : @require_role(2) + @require_permission('can_admin_portal') (CRUD).
          Le /check est ouvert aux utilisateurs (lecture d'etat, machine validee).
  - A03 : requetes parametrees ; jours/heures valides cote serveur.
"""
import logging
import re
from datetime import datetime

from flask import Blueprint, jsonify, request

from routes.helpers import (
    require_api_key, require_role, require_permission, require_machine_access,
    threaded_route, get_db_connection, get_current_user, logger,
)
import maintenance as mw

bp = Blueprint('maintenance', __name__)

_TIME_RE = re.compile(r'^([01]?\d|2[0-3]):[0-5]\d$')


def _clean_days(raw):
    """Normalise une liste/CSV de jours 0-6 en CSV trie sans doublon."""
    if isinstance(raw, str):
        raw = raw.split(',')
    days = sorted({int(x) for x in raw if str(x).strip().isdigit() and 0 <= int(x) <= 6})
    return ','.join(str(d) for d in days)


@bp.route('/maintenance/windows', methods=['GET'])
@require_api_key
@require_role(2)
@require_permission('can_admin_portal')
@threaded_route
def list_windows():
    """Liste toutes les fenetres (avec nom de machine pour le scope machine).

    Rend aussi, PAR FENETRE, `active_now` : le verdict calcule par `_in_window`,
    c'est-a-dire par la fonction meme qui bloque les actions mutantes.

    ══ POURQUOI CE CHAMP EXISTE ════════════════════════════════════════════════

    L'ancien portail calculait ce verdict DANS LE NAVIGATEUR
    (`legacy/maintenance/js/main.js:26-35`), sur l'horloge du navigateur. Or
    l'enforcement se fait ici, sur l'horloge de CE conteneur. Mesure du
    2026-08-25 : navigateur en CEST, conteneur en UTC, **deux heures d'ecart**.
    Une fenetre 22:00 -> 06:00 saisie en heure locale etait donc annoncee
    « active » de 22:00 a 06:00 locales alors qu'elle n'autorisait reellement que
    de 00:00 a 08:00 locales. Deux bandes de deux heures ou la page et
    l'application se contredisaient, dans les deux sens — et le refus se produit
    sur une AUTRE page, donc rien ne rapprochait les deux.

    Le verdict part desormais d'ici. On ne deplace pas la regle vers le
    navigateur : on l'y annonce, telle qu'elle sera appliquee.

    `server_time` accompagne la liste pour que la page puisse NOMMER l'horloge
    employee. Sans elle, un exploitant voyant « fermee » a 07:00 n'a aucun moyen
    de comprendre pourquoi.
    """
    maintenant = datetime.now()
    jour = maintenant.weekday()

    with get_db_connection() as conn:
        cur = conn.cursor(dictionary=True)
        cur.execute(
            "SELECT w.*, m.name AS machine_name FROM maintenance_windows w "
            "LEFT JOIN machines m ON w.machine_id = m.id ORDER BY w.scope, w.name")
        rows = cur.fetchall()
        for r in rows:
            debut = mw._to_time(r.get('start_time'))
            fin = mw._to_time(r.get('end_time'))
            # Une fenetre desactivee n'est jamais active : `is_allowed` ne la lit
            # meme pas (`WHERE enabled = 1`). Le dire ici evite que la page ait a
            # recombiner deux champs pour retrouver la meme conclusion.
            r['active_now'] = bool(
                r.get('enabled') and debut is not None and fin is not None
                and mw._in_window(maintenant, jour, mw._parse_days(r.get('days')), debut, fin))
            for k in ('start_time', 'end_time'):
                tv = mw._to_time(r.get(k))
                r[k] = tv.strftime('%H:%M') if tv else None
            if r.get('created_at') and hasattr(r['created_at'], 'isoformat'):
                r['created_at'] = r['created_at'].isoformat()
    return jsonify({
        'success': True,
        'windows': rows,
        # L'horloge QUI DECIDE, telle quelle. `%Z` est vide sur un conteneur en
        # UTC sans tzdata configure : on ajoute donc le decalage, qui ne l'est
        # jamais.
        'server_time': maintenant.strftime('%H:%M'),
        'server_offset': maintenant.astimezone().strftime('%z'),
    })


@bp.route('/maintenance/windows', methods=['POST'])
@require_api_key
@require_role(2)
@require_permission('can_admin_portal')
@threaded_route
def create_window():
    """Cree une fenetre. Body : {name, scope, machine_id?, days[], start_time, end_time, enabled}."""
    data = request.get_json(silent=True) or {}
    name = (data.get('name') or '').strip()
    if not name:
        return jsonify({'success': False, 'message': 'Nom requis'}), 400
    scope = data.get('scope', 'global')
    if scope not in ('global', 'machine'):
        scope = 'global'
    machine_id = data.get('machine_id') if scope == 'machine' else None
    try:
        machine_id = int(machine_id) if machine_id is not None else None
    except (ValueError, TypeError):
        return jsonify({'success': False, 'message': 'machine_id invalide'}), 400
    if scope == 'machine' and machine_id is None:
        return jsonify({'success': False, 'message': 'machine_id requis pour le scope machine'}), 400

    days = _clean_days(data.get('days') or [])
    if not days:
        return jsonify({'success': False, 'message': 'Au moins un jour requis'}), 400
    start_t = (data.get('start_time') or '').strip()
    end_t = (data.get('end_time') or '').strip()
    if not _TIME_RE.match(start_t) or not _TIME_RE.match(end_t):
        return jsonify({'success': False, 'message': 'Heures invalides (HH:MM)'}), 400
    enabled = 1 if data.get('enabled', True) else 0
    user_id, _ = get_current_user()

    with get_db_connection() as conn:
        cur = conn.cursor()
        cur.execute(
            "INSERT INTO maintenance_windows (name, scope, machine_id, days, start_time, end_time, enabled, created_by) "
            "VALUES (%s,%s,%s,%s,%s,%s,%s,%s)",
            (name[:100], scope, machine_id, days, start_t + ':00', end_t + ':00', enabled, user_id))
        conn.commit()
        wid = cur.lastrowid
    return jsonify({'success': True, 'id': wid})


@bp.route('/maintenance/windows/<int:window_id>', methods=['PUT'])
@require_api_key
@require_role(2)
@require_permission('can_admin_portal')
@threaded_route
def update_window(window_id):
    """Met a jour une fenetre (champs partiels)."""
    data = request.get_json(silent=True) or {}
    sets, vals = [], []
    if 'name' in data and (data['name'] or '').strip():
        sets.append('name = %s')
        vals.append(data['name'].strip()[:100])
    if 'days' in data:
        d = _clean_days(data['days'] or [])
        if d:
            sets.append('days = %s')
            vals.append(d)
    for col in ('start_time', 'end_time'):
        if col in data:
            v = (data[col] or '').strip()
            if not _TIME_RE.match(v):
                return jsonify({'success': False, 'message': f'{col} invalide (HH:MM)'}), 400
            sets.append(f'{col} = %s')
            vals.append(v + ':00')
    if 'enabled' in data:
        sets.append('enabled = %s')
        vals.append(1 if data['enabled'] else 0)
    if not sets:
        return jsonify({'success': True})
    with get_db_connection() as conn:
        cur = conn.cursor()
        vals.append(window_id)
        cur.execute(f"UPDATE maintenance_windows SET {', '.join(sets)} WHERE id = %s", vals)
        conn.commit()
    return jsonify({'success': True})


@bp.route('/maintenance/windows/<int:window_id>', methods=['DELETE'])
@require_api_key
@require_role(2)
@require_permission('can_admin_portal')
@threaded_route
def delete_window(window_id):
    """Supprime une fenetre."""
    with get_db_connection() as conn:
        cur = conn.cursor()
        cur.execute("DELETE FROM maintenance_windows WHERE id = %s", (window_id,))
        conn.commit()
        deleted = cur.rowcount > 0
    return jsonify({'success': True, 'deleted': deleted})


@bp.route('/maintenance/check', methods=['GET'])
@require_api_key
@require_machine_access
@threaded_route
def check_window():
    """Indique si une action mutante est autorisee maintenant sur une machine.
    Query : ?machine_id=<int>. Utilise par le frontend pour prevenir l'operateur."""
    from ssh_utils import validate_machine_id
    try:
        machine_id = validate_machine_id(request.args.get('machine_id'))
    except (ValueError, TypeError):
        return jsonify({'success': False, 'message': 'machine_id invalide'}), 400
    _uid, role = get_current_user()
    allowed, reason = mw.is_allowed(machine_id, role=role)
    return jsonify({'success': True, 'allowed': allowed, 'reason': reason})

File: laravel/public/js/maintenance.js
/**
 * maintenance.js - Fenetres de maintenance.
 *
 * Quatre differences avec `legacy/maintenance/js/main.js`, toutes des
 * corrections :
 *
 *   1. **AUCUNE BOITE NATIVE.** Le legacy supprime derriere un `confirm()` :
 *      cette boite recouvre la ligne sur laquelle on decide, ne se style pas, et
 *      BLOQUE Puppeteer. Ici la confirmation s'ouvre EN LIGNE.
 *   2. **Rendu par `textContent`**, jamais par interpolation.
 *   3. **Un echec RESEAU se voit** : chaque appel est enveloppe.
 *   4. **Le formulaire se cache par `hidden`**, pas par une classe : le projet a
 *      paye qu'une regle `display:` sur la classe rendait `hidden` sans effet.
 *
 * ══ LA CINQUIEME, ET LA PLUS IMPORTANTE : LE VERDICT N'EST PLUS RECALCULE ═══
 *
 * « Cette fenetre est-elle ouverte maintenant ? » etait calcule DEUX FOIS : ici
 * en JavaScript, et dans `backend/maintenance.py:_in_window` en Python. Le
 * premier jet de ce fichier faisait de meme, en promettant de « suivre le Python
 * pas a pas ».
 *
 * C'etait la mauvaise reponse, et la mesure du 2026-08-25 l'a montre : les deux
 * calculs ne lisent pas la MEME HORLOGE. Le navigateur est en CEST, le conteneur
 * qui applique la regle est en UTC — deux heures d'ecart. Une fenetre
 * 22:00 -> 06:00 etait annoncee « active » de 22:00 a 06:00 locales alors qu'elle
 * n'autorisait reellement que de 00:00 a 08:00 locales. Suivre le Python « pas a
 * pas » ne protege de rien quand ce n'est pas le pas qui differe, mais l'heure.
 *
 * `list_windows` rend donc desormais `active_now` par fenetre, calcule par
 * `_in_window` elle-meme, et `server_time`. Ce fichier ne fait plus que
 * l'AFFICHER. La regle n'est pas deplacee vers le navigateur : elle y est
 * annoncee, telle qu'elle sera appliquee.
 *
 * L'heure du serveur est montree a cote du tableau, sans quoi un exploitant
 * voyant « fermee » a 07:00 n'aurait aucun moyen de comprendre pourquoi.
 */
(function () {
    'use strict';

    const PASSERELLE = '/api/gateway';
    const L = JSON.parse(document.getElementById('maint-libelles').textContent);
    const JOURS = [L.mon, L.tue, L.wed, L.thu, L.fri, L.sat, L.sun];
    const corps = document.querySelector('[data-rw="maint-corps"]');
    const formulaire = document.querySelector('[data-rw="maint-formulaire"]');
    const message = document.querySelector('[data-rw="maint-message"]');

    function jetonCsrf() {
        const m = document.querySelector('meta[name="csrf-token"]');

        return m ? m.content : '';
    }

    function annonce(texte, type) {
        message.textContent = texte;
        message.className = 'rw-annonce' + (type ? ' rw-annonce--' + type : '');
    }

    async function appelle(chemin, options) {
        const parametres = Object.assign({
            headers: { 'X-Requested-With': 'XMLHttpRequest' },
            credentials: 'same-origin',
        }, options || {});
        if (parametres.method && parametres.method !== 'GET') {
            parametres.headers = Object.assign({}, parametres.headers, {
                'X-CSRF-TOKEN': jetonCsrf(),
                'Content-Type': 'application/json',
            });
        }
        try {
            const r = await fetch(PASSERELLE + chemin, parametres);
            let json = null;
            try { json = await r.json(); } catch (e) { /* reponse non JSON */ }

            return { ok: r.ok, statut: r.status, corps: json };
        } catch (e) {
            return { ok: false, statut: 0, corps: null, reseau: true };
        }
    }

    function cellule(texte) {
        const td = document.createElement('td');
        td.textContent = texte == null ? '' : String(texte);

        return td;
    }

    function pastillesJours(csv) {
        const td = document.createElement('td');
        const set = String(csv || '').split(',').map((s) => parseInt(s, 10));
        JOURS.forEach((nom, i) => {
            const p = document.createElement('span');
            p.className = 'rw-pastille ' + (set.includes(i) ? 'rw-pastille--info' : 'rw-pastille--neutre');
            p.textContent = nom;
            td.appendChild(p);
        });

        return td;
    }

    function pastilleEtat(f) {
        const td = document.createElement('td');
        const p = document.createElement('span');
        if (! f.enabled) {
            p.className = 'rw-pastille rw-pastille--neutre';
            p.textContent = L.disabled;
        } else if (f.active_now) {
            p.className = 'rw-pastille rw-pastille--ok';
            p.textContent = L.active_now;
        } else {
            p.className = 'rw-pastille rw-pastille--attente';
            p.textContent = L.closed_now;
        }
        td.appendChild(p);

        return td;
    }

    /** La confirmation, EN LIGNE et sous la ligne concernee. */
    function ouvreConfirmation(tr, f) {
        if (tr.nextElementSibling?.dataset.rw === 'maint-panneau') return;

        const ligne = document.createElement('tr');
        ligne.dataset.rw = 'maint-panneau';
        const td = document.createElement('td');
        td.colSpan = 6;
        /*
         * LE CONTENEUR FLEX VA *DANS* LA CELLULE, JAMAIS SUR ELLE.
         *
         * `.rw-panneau-decision` porte `display: flex`. Pose sur un `<td>`, il
         * ecrase `display: table-cell` : la cellule SORT du modele de tableau et
         * son `colspan` est ignore — le panneau s'arretait au tiers de la largeur
         * sur un ecran large, le reste de la ligne restant blanc. Aucune
         * assertion DOM ne pouvait le voir, `colSpan` valant bien 6.
         *
         * Defaut vu a l'image sur `graylog/` le 2026-08-26, puis retrouve ici et
         * dans trois autres fichiers du portage. Voir PARITE E-139.
         */
        const cadre = document.createElement('div');
        cadre.className = 'rw-panneau-decision';

        const bloc = document.createElement('div');
        bloc.className = 'rw-panneau-decision__texte';
        const titre = document.createElement('strong');
        titre.textContent = L.confirm_titre;
        const aide = document.createElement('p');
        aide.className = 'rw-aide';
        aide.textContent = String(L.confirm_aide).replace(':name', f.name);
        bloc.append(titre, aide);

        const actions = document.createElement('div');
        actions.className = 'rw-panneau-decision__actions';

        const annuler = document.createElement('button');
        annuler.type = 'button';
        annuler.className = 'rw-bouton rw-bouton--discret';
        annuler.dataset.rw = 'maint-annuler-suppression';
        annuler.textContent = L.confirm_annuler;
        annuler.addEventListener('click', () => ligne.remove());

        const confirmer = document.createElement('button');
        confirmer.type = 'button';
        confirmer.className = 'rw-bouton rw-bouton--danger';
        confirmer.dataset.rw = 'maint-confirmer';
        confirmer.textContent = L.confirm_supprimer;
        confirmer.addEventListener('click', () => supprime(f, confirmer, ligne));

        actions.append(annuler, confirmer);
        cadre.append(bloc, actions);
        td.appendChild(cadre);
        ligne.appendChild(td);
        tr.after(ligne);
    }

    function rend(fenetres) {
        corps.textContent = '';
        if (! fenetres.length) {
            const tr = document.createElement('tr');
            const td = cellule(L.empty);
            td.colSpan = 6;
            td.className = 'rw-vide';
            tr.appendChild(td);
            corps.appendChild(tr);

            return;
        }
        fenetres.forEach((f) => {
            const tr = document.createElement('tr');
            tr.appendChild(cellule(f.name));
            tr.appendChild(cellule(f.scope === 'machine'
                ? (f.machine_name || ('#' + f.machine_id))
                : L.scope_global));
            tr.appendChild(pastillesJours(f.days));
            tr.appendChild(cellule(f.start_time + ' → ' + f.end_time));
            tr.appendChild(pastilleEtat(f));

            const td = document.createElement('td');
            const basculer = document.createElement('button');
            basculer.type = 'button';
            basculer.className = 'rw-bouton rw-bouton--discret rw-bouton--minuscule';
            basculer.dataset.rw = 'maint-basculer';
            basculer.textContent = f.enabled ? L.disable : L.enable;
            basculer.addEventListener('click', () => bascule(f, basculer));
            const supprimer = document.createElement('button');
            supprimer.type = 'button';
            supprimer.className = 'rw-bouton rw-bouton--discret rw-bouton--minuscule';
            supprimer.dataset.rw = 'maint-supprimer';
            supprimer.textContent = L.delete;
            supprimer.addEventListener('click', () => ouvreConfirmation(tr, f));
            td.append(basculer, supprimer);
            tr.appendChild(td);
            corps.appendChild(tr);
        });
    }

    async function charge() {
        const r = await appelle('/maintenance/windows');
        if (! r.ok || ! r.corps || ! r.corps.success) {
            annonce(r.reseau ? L.err_reseau : L.err_load, 'echec');
            rend([]);

            return;
        }
        rend(r.corps.windows || []);
        montreHorlogeServeur(r.corps.server_time, r.corps.server_offset);
        majEtatPortee(r.corps.windows || []);
    }

    /**
     * LA PASTILLE D'ENSEMBLE DOIT SUIVRE LES GESTES DE LA PAGE.
     *
     * Elle est rendue une premiere fois par le serveur, qui a compte — c'est ce
     * qui evite un affichage vide au chargement. Mais la page CREE, BASCULE et
     * SUPPRIME des fenetres : sans cette mise a jour, elle continue d'annoncer
     * l'etat d'avant. Defaut mesure le 2026-08-25 : apres avoir cree une fenetre
     * limitee a une machine, la pastille affichait encore « Aucune restriction ».
     *
     * Le calcul se fait sur la liste que le backend vient de rendre, et il ne
     * duplique aucune regle d'horaire : ce sont deux comptages sur `scope` et
     * `enabled`, exactement la condition du `WHERE` de `is_allowed`.
     */
    function majEtatPortee(fenetres) {
        const pastille = document.querySelector('[data-rw="maint-etat-flotte"]');
        if (! pastille) return;

        const globales = fenetres.filter((f) => f.enabled && f.scope === 'global').length;
        const machines = new Set(fenetres
            .filter((f) => f.enabled && f.scope === 'machine' && f.machine_id != null)
            .map((f) => f.machine_id)).size;

        let etat = 'libre';
        let libelle = L.etat_libre;
        if (globales > 0) {
            etat = 'flotte';
            libelle = L.etat_restreint;
        } else if (machines > 0) {
            etat = 'machines';
            libelle = pluriel(L.etat_machines, machines).replace(':n', String(machines));
        }

        pastille.setAttribute('data-rw-etat', etat);
        pastille.className = 'rw-pastille ' + (etat === 'libre' ? 'rw-pastille--ok' : 'rw-pastille--attente');
        pastille.textContent = libelle;
        pastille.title = (L.etat_detail || '')
            .replace(':n', String(fenetres.filter((f) => f.enabled).length))
            .replace(':g', String(globales));
    }

    /**
     * Les formes de pluriel de Laravel arrivent ici telles quelles, separees par
     * une barre : « :n machine restreinte|:n machines restreintes ». Le cadre
     * choisit d'habitude la branche cote PHP ; ce fichier n'a que la chaine, il
     * doit donc trancher lui-meme. Deux formes suffisent aux deux langues du
     * projet — une troisieme se lirait ici comme une absence de branche, pas
     * comme un silence.
     */
    function pluriel(forme, nombre) {
        const formes = String(forme || '').split('|');

        return formes.length > 1 && nombre !== 1 ? formes[1] : formes[0];
    }

    /**
     * NOMMER L'HORLOGE QUI DECIDE.
     *
     * Le verdict « active maintenant » vient du serveur. Si son horloge n'est pas
     * celle du navigateur — mesure du 2026-08-25 : UTC contre CEST — un
     * exploitant lisant « fermee » a 07:00 doit pouvoir voir POURQUOI, sans quoi
     * il conclura a une panne. On ne l'affiche que lorsque les deux diffèrent :
     * une information toujours presente cesse d'etre lue.
     */
    function montreHorlogeServeur(heure, decalage) {
        const cible = document.querySelector('[data-rw="maint-horloge"]');
        if (! cible) return;
        if (! heure) { cible.hidden = true; return; }
        const d = new Date();
        const local = String(d.getHours()).padStart(2, '0') + ':'
            + String(d.getMinutes()).padStart(2, '0');
        // Comparaison a la MINUTE, et une minute de tolerance : deux horloges
        // justes peuvent se croiser au changement de minute, et clignoter alors
        // ferait douter d'un ecart qui n'existe pas.
        const differe = Math.abs(enMinutesHM(local) - enMinutesHM(heure)) > 1;
        cible.textContent = L.horloge_serveur
            .replace(':heure', heure)
            .replace(':decalage', decalage || '');
        cible.hidden = ! differe;
    }

    function enMinutesHM(hhmm) {
        const p = String(hhmm || '0:0').split(':');

        return (Number(p[0]) * 60) + (Number(p[1]) || 0);
    }

    async function enregistre(bouton) {
        const nom = document.querySelector('[data-rw="maint-nom"]').value.trim();
        if (! nom) { annonce(L.err_name, 'echec'); return; }
        const portee = document.querySelector('[data-rw="maint-portee"]').value;
        const jours = Array.from(document.querySelectorAll('[data-rw="maint-jour"]'))
            .filter((c) => c.checked)
            .map((c) => parseInt(c.value, 10));
        if (! jours.length) { annonce(L.err_days, 'echec'); return; }

        bouton.disabled = true;
        const r = await appelle('/maintenance/windows', {
            method: 'POST',
            body: JSON.stringify({
                name: nom,
                scope: portee,
                machine_id: portee === 'machine'
                    ? parseInt(document.querySelector('[data-rw="maint-machine"]').value, 10)
                    : null,
                days: jours,
                start_time: document.querySelector('[data-rw="maint-debut"]').value,
                end_time: document.querySelector('[data-rw="maint-fin"]').value,
                enabled: document.querySelector('[data-rw="maint-activee"]').checked,
            }),
        });
        bouton.disabled = false;
        if (r.ok && r.corps && r.corps.success) {
            annonce(L.saved, 'ok');
            ferme();
            charge();

            return;
        }
        annonce(r.reseau ? L.err_reseau : ((r.corps && r.corps.message) || L.err_save), 'echec');
    }

    async function bascule(f, bouton) {
        bouton.disabled = true;
        const r = await appelle('/maintenance/windows/' + encodeURIComponent(f.id), {
            method: 'PUT',
            body: JSON.stringify({ enabled: ! f.enabled }),
        });
        bouton.disabled = false;
        if (r.ok && r.corps && r.corps.success) { charge(); return; }
        annonce(r.reseau ? L.err_reseau : L.err_save, 'echec');
    }

    async function supprime(f, bouton, ligne) {
        bouton.disabled = true;
        const r = await appelle('/maintenance/windows/' + encodeURIComponent(f.id),
            { method: 'DELETE' });
        bouton.disabled = false;
        ligne.remove();
        if (r.ok && r.corps && r.corps.success) {
            annonce(L.deleted, 'ok');
            charge();

            return;
        }
        annonce(r.reseau ? L.err_reseau : L.err_save, 'echec');
    }

    function ouvre() { formulaire.hidden = false; }
    function ferme() {
        formulaire.hidden = true;
        document.querySelector('[data-rw="maint-nom"]').value = '';
    }

    /** Le choix de machine n'a de sens que si la portee est `machine`. */
    function ajusteMachine() {
        const bloc = document.querySelector('[data-rw="maint-machine-bloc"]');
        bloc.hidden = document.querySelector('[data-rw="maint-portee"]').value !== 'machine';
    }

    document.querySelector('[data-rw="maint-nouvelle"]').addEventListener('click', ouvre);
    document.querySelector('[data-rw="maint-annuler"]').addEventListener('click', ferme);
    document.querySelector('[data-rw="maint-portee"]').addEventListener('change', ajusteMachine);
    document.querySelector('[data-rw="maint-enregistrer"]')
        .addEventListener('click', (e) => enregistre(e.currentTarget));
    ajusteMachine();
    charge();
})();
